imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,334 @@
+
+    Notes
+    -----
+    The structure of this function goes as follows:
+    1. Initialize the HealpixSkyMap object with the specified properties.
+    2. Iterate over the input pointing sets and read them into UltraPointingSet objects.
+    3. For each pointing set, weight certain variables by exposure and solid angle of
+    the pointing set pixels.
+    4. Project the pointing set values to the map using the push method.
+    5. Perform subsequent processing for weighted quantities at the SkyMap level
+    (e.g., divide weighted quantities by their summed weights to
+    get their weighted mean)
+    6. Calculate corrected count rate with background subtraction applied.
+    7. Calculate flux and flux uncertainty.
+    8. Drop unnecessary variables from the map.
+    """
+    if output_map_properties.sky_tiling_type is ena_maps.SkyTilingType.HEALPIX:
+        map_nside, map_nested = (
+            output_map_properties.nside,
+            output_map_properties.nested,
+        )
+    else:
+        map_nside, map_nested = (DEFAULT_L2_HEALPIX_NSIDE, DEFAULT_L2_HEALPIX_NESTED)
+
+    # Initialize the HealpixSkyMap object
+    skymap = ena_maps.HealpixSkyMap(
+        nside=map_nside,
+        nested=map_nested,
+        spice_frame=output_map_properties.spice_reference_frame,
+    )
+
+    # Add additional data variables to the map
+    output_map_properties.values_to_push_project.extend(
+        [
+            "observation_time",
+            "pointing_set_exposure_times_solid_angle",
+            "num_pointing_set_pixel_members",
+        ]
+    )
+
+    for ultra_l1c_pset in ultra_l1c_psets:
+        pointing_set = read_into_pointing_set(ultra_l1c_pset)
+        logger.info(f"PSET epoch: {pointing_set.epoch}")
+
+        pointing_set.data["num_pointing_set_pixel_members"] = xr.DataArray(
+            np.ones(pointing_set.num_points, dtype=int),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        pointing_set.data["observation_time"] = xr.DataArray(
+            np.full(pointing_set.num_points, pointing_set.epoch),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        # Add solid_angle * exposure of pointing set as data_var
+        # so this quantity is projected to map pixels for use in weighted averaging
+        pointing_set.data["pointing_set_exposure_times_solid_angle"] = (
+            pointing_set.data["exposure_factor"] * pointing_set.solid_angle
+        )
+
+        # Initial processing for weighted quantities at PSET level
+        # Weight the values by exposure and solid angle
+        # (in that order to avoid double weighting by solid angle)
+        for (
+            quantity_to_weight
+        ) in VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE:
+            pointing_set.data[quantity_to_weight] = (
+                pointing_set.data[quantity_to_weight]
+                * pointing_set.data["pointing_set_exposure_times_solid_angle"]
+            )
```

Comment:
I think you could use `*=` here?

Or alternatively could you do this as a one-liner instead of a for-loop by passing in the list of variables?

`pointing_set[VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE] *= pointing_set["pointing_set_exposure_times_solid_angle"]`

Reply:
Sometimes in manual testing I've have weird results with `*=` and `/=` on xarray objects, but I think you're right that it works fine here. I didn't realize you could use the latter syntax, but that's great! I'm switching to that.